spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala

```diff
@@ -0,0 +1,333 @@
+import com.intel.analytics.bigdl.Module
+import com.intel.analytics.bigdl.nn.Graph.ModuleNode
+import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity, TensorModule}
+import com.intel.analytics.bigdl.tensor.{Storage, Tensor}
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.T
+
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+
+/**
+ * Transformer model from "Attention Is All You Need".
+ * The Transformer model consists of an encoder and a decoder. Both are stacks
+ * of self-attention layers followed by feed-forward layers. This model yields
+ * good results on a number of problems, especially in NLP and machine translation.
+ * See "Attention Is All You Need" (https://arxiv.org/abs/1706.03762) for the full
+ * description of the model and the results obtained with its early version.
+ * @param hiddenSize
+ * @param numHeads
+ * @param filterSize
+ * @param numHiddenlayers
+ * @param postprocessDropout
+ * @param attentionDropout
+ * @param reluDropout
+ * @tparam T The numeric type in this module parameters.
+ */
+class TransformerLayer[T: ClassTag](
+   val vocabSize: Int,
+   val hiddenSize: Int,
+   val numHeads: Int,
+   val filterSize: Int,
+   val numHiddenlayers: Int,
+   val postprocessDropout: Float,
+   val attentionDropout: Float,
+   val reluDropout: Float,
+   val problem: ProblemType = LanguageModel)
+  (implicit ev: TensorNumeric[T]) extends BaseModule[T] {
+
+  override def buildModel(): Module[T] = {
+    problem match {
+      case LanguageModel => buildLM()
+      case Translation => buildTranslation()
+    }
+  }
+
+  // inputs: int tensor with shape [batch_size, input_length].
+  // targets: None or int tensor with shape [batch_size, target_length].
+  private def buildTranslation(): Module[T] = {
+    val inNode = Input()
+    val tarNode = Input()
+    val attention_bias = new AttentionBiasConstant().inputs(inNode)
+
+    val embedding = LookupTable[T](vocabSize, hiddenSize)
+    val embeddingDupicate = LookupTable[T](vocabSize, hiddenSize)
+    // parameter share
+    val params1 = embedding.getParameters()
+    val params2 = embeddingDupicate.getParameters()
+    params1._1.set(params2._1)
+    params1._2.set(params2._2)
+
+    val constantValue = math.sqrt(hiddenSize)
+    val embeddingInput = MulConstant(constantValue).inputs(embedding.inputs(inNode))
+    val embeddingOutput = MulConstant(constantValue).inputs(embeddingDupicate.inputs(tarNode))
+    val encoder_outputs = encode(embeddingInput, attention_bias)
+    val outNode = decode(embeddingOutput, encoder_outputs, attention_bias)
+    Graph(Array(inNode, tarNode), outNode)
+  }
+
+  private def buildLM(): Module[T] = {
+    val inNode = Input()
+    val constantValue = math.sqrt(hiddenSize)
+    val embeddingInput = MulConstant(constantValue).inputs(
+      LookupTable[T](vocabSize, hiddenSize).inputs(inNode))
+    val outNode = decode(embeddingInput)
```

Comment:
outputNode ? we I assume we name similar in other code as well

Reply:
ok, changed